### PR TITLE
feat: add `physicalFilename` property to the rule context

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Refers #16999 

This PR creates `context.physicalFilename` and deprecates context.getPhysicalFilename().

#### Is there anything you'd like reviewers to focus on?

Did I miss any references?

### CODE REVIEW

File: docs/src/extend/custom-rules.md [hunk 1]
```diff
@@ -145,7 +145,8 @@ Additionally, the `context` object has the following methods:
     * If the node is an `ImportSpecifier`, `ImportDefaultSpecifier`, or `ImportNamespaceSpecifier`, the declared variable is returned.
     * Otherwise, if the node does not declare any variables, an empty array is returned.
 * `getFilename()`: Returns the filename associated with the source.
-* `getPhysicalFilename()`: When linting a file, it returns the full path of the file on disk without any code block information. When linting text, it returns the value passed to `—stdin-filename` or `<text>` if not specified.
+* `getPhysicalFilename()`: (**Deprecated:** Use `context#physicalFilename` instead.) When linting a file, it returns the full path of the file on disk without any code block information. When linting text, it returns the value passed to `—stdin-filename` or `<text>` if not specified.
```
Comment (mdjermanovic) [inline docs/src/extend/custom-rules.md:148]:
```suggestion
* `getPhysicalFilename()`: (**Deprecated:** Use `context.physicalFilename` instead.) When linting a file, it returns the full path of the file on disk without any code block information. When linting text, it returns the value passed to `—stdin-filename` or `<text>` if not specified.
```

File: tests/lib/linter/linter.js [hunk 2]
```diff
@@ -1846,7 +1846,7 @@ describe("Linter", () => {
             linter.defineRule(code, {
                 create: context => ({
                     Literal(node) {
-                        context.report(node, context.getPhysicalFilename());
+                        context.report(node, context.physicalFilename);
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:1849]: Similar to https://github.com/eslint/eslint/pull/17108#discussion_r1176205663, I think we should keep testing `context.getPhysicalFilename()`.

```suggestion
                        assert.strictEqual(context.getPhysicalFilename(), context.physicalFilename);
                        context.report(node, context.physicalFilename);
```

(the same for other places where we switched to testing `context.physicalFilename`)

Comment (snitin315) [reply tests/lib/linter/linter.js:1849]: We do have dedicated test cases for `context.getPhysicalFilename()`, do we still need this here?

https://github.com/eslint/eslint/blob/1aaece943ba73987344fcd8941109d4484335347/tests/lib/linter/linter.js#L9112

Comment (mdjermanovic) [reply tests/lib/linter/linter.js:1849]: It's only one test and it's under "Linter with FlatConfigArray", so we wouldn't have any tests with `context.getPhysicalFilename()` in eslintrc mode, which is what's currently used in production.

Comment (snitin315) [reply tests/lib/linter/linter.js:1919]: @mdjermanovic Updated 👍🏻

File: docs/src/extend/custom-rules.md [hunk 3]
```diff
@@ -124,6 +124,7 @@ As the name implies, the `context` object contains information that is relevant
 The `context` object has the following properties:
 
 * `id`: (`string`) The rule ID.
+* `physicalFilename`: (`string`) When linting a file, it returns the full path of the file on disk without any code block information. When linting text, it returns the value passed to `—stdin-filename` or `<text>` if not specified.
```
Comment (mdjermanovic) [inline docs/src/extend/custom-rules.md:127]: If possible, it would be nice to reword this to avoid using the word "returns" since this is not a method.

**Review by mdjermanovic — APPROVED** [review]

File: docs/src/extend/custom-rules.md (hunk 3 above)
Comment (snitin315) [reply docs/src/extend/custom-rules.md:127]: @mdjermanovic How about this?

```suggestion
* `physicalFilename`: (`string`) When linting a file, it provides the full path of the file on disk without any code block information. When linting text, it provides the value passed to `—stdin-filename` or `<text>` if not specified.
```

Comment (mdjermanovic) [reply docs/src/extend/custom-rules.md:127]: Sounds good to me!

Comment (snitin315) [reply docs/src/extend/custom-rules.md]: @mdjermanovic Updated 👍🏻

File: tests/lib/linter/linter.js [hunk 4]
```diff
@@ -6757,7 +6761,7 @@ var a = "test2";
                 create: context => ({
                     Program(ast) {
                         receivedFilenames.push(context.filename);
-                        receivedPhysicalFilenames.push(context.getPhysicalFilename());
+                        receivedPhysicalFilenames.push(context.physicalFilename);
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:6764]: It would be good to add `assert.strictEqual(context.getFilename(), context.filename)` and `assert.strictEqual(context.getPhysicalFilename(), context.physicalFilename)` here.

File: tests/lib/linter/linter.js [hunk 5]
```diff
@@ -15523,7 +15565,7 @@ var a = "test2";
                                 return {
                                     Program(ast) {
                                         receivedFilenames.push(context.filename);
-                                        receivedPhysicalFilenames.push(context.getPhysicalFilename());
+                                        receivedPhysicalFilenames.push(context.physicalFilename);
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:15568]: Same here.

File: tests/lib/linter/linter.js (hunk 4 above)
Comment (snitin315) [reply tests/lib/linter/linter.js:6767]: @mdjermanovic Added 👍🏻

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
